#! /bin/sh -e

if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch $pdir -f --no-backup-if-mismatch -p0 < $0;;
    -unpatch) patch $pdir -f --no-backup-if-mismatch -R -p0 < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

From: Satoshi Koike <s-koike@jaist.ac.jp>
Sender: Satoshi Koike <s-koike@scene.jaist.ac.jp>
To: submit@bugs.debian.org
Subject: Bug#155436: bash: Completion by TAB in quoted sequence(`...`) with Multibyte Locale
Date: Mon, 05 Aug 2002 06:29:19 +0900

Package: bash
Version: 2.05b-2
Severity: normal

# DP: Completion by TAB in quoted sequence(`...`) with Multibyte Locale

I use bash with Japanese Locale (ja_JP.eucJP). When I use the command
completion by [TAB] key in quoted sequence using `...`, bash freeze
and I must push Ctrl-C for getting bash prompt.

The followed is example.

> $ echo `pwd[TAB]
[bash is freezing and pushing Ctrl-C will get prompt]

I checked the source code, and I knew that bash stopped in function
unclosed_pair(char *, int, char *) in subst.c. unclosed_pair() uses
the macro ADVANCE_CHAR defined in include/shmbutil.h. This macro
increases the pointer variable, but it doesn't increase in the
multibyte environment. Because in the multibyte environment, it
increases the pointer using the return value of mbrlen(3). mbrlen()
returns ZERO when the given word is the termination.

So, I think that the macro ADVANCED_CHAR will be change like the
followed patch.

------------------------------ patch ------------------------------
--- include/shmbutil.h.orig	2002-08-05 06:26:13.000000000 +0900
+++ include/shmbutil.h	2002-08-05 06:19:37.000000000 +0900
@@ -120,6 +120,8 @@
 		state = state_bak; \
 		(_i)++; \
 	      } \
+            else if (mblength == 0) \
+              (_i)++; \
 	    else \
 	      (_i) += mblength; \
 	  } \
------------------------------ patch ------------------------------
